File: coinChange2BF.java
// Time Complexity : O(2^(m+n))
// Space Complexity : O(m + n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach
class Solution {
    public int change(int amount, int[] coins) {
        //base case in case the array is empty
        if(coins == null || coins.length == 0)
            return 0;
        
        //index is 0 to begin with
        return helper(coins, amount, 0);
    }
    
    private int helper(int[] coins, int amount, int index){
        //base case: if amount is 0 it means we have found a path
        if(amount == 0)
            return 1;
        //base case: if index is out of bounds or amount becomes less than 0, it is not a valid path
        if(amount < 0 || index >= coins.length)
            return 0;
        
        //logic
        //choose the coin at the index
        int case1 = helper(coins, amount - coins[index], index);
            
        //dont choose that denomination
        int case2 = helper(coins, amount, index + 1);
        
        return(case1 + case2);
            
    }
}

File: Coin_Change2_BF.py
# We are following the same path as the coin change1 problem
# using a BruteForce Recursion method
# Complexity is exponential 2^n
# This will fail for ceratin cases as time limit excedded.
# Since we have repeated sub problems as seen in the Recursion Tree
# Lets use that as a tell tale sign of Dynamic Programming problem.


class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        if coins == None:
            return 0
        return self.helper(amount, coins, idx=0)

    def helper(self,amount, coins,idx ):
        # base case
        if amount == 0:return 1
        if amount < 0 or idx == len(coins):return 0
        # logic
        pick = self.helper(amount-coins[idx],coins, idx)
        skip = self.helper(amount, coins, idx +1)
        return pick + skip
